rust/xaynet-server/src/storage/api.rs

```diff
@@ -0,0 +1,492 @@
+    /// - RedisError
+    ///
+    /// ## Protocol
+    /// None
+    async fn get_seed_dict(&self) -> StorageResult<SeedDict>;
+
+    /////////// Mask dict
+
+    /// Increments the mask score with the given [`MaskObject`].
+    /// The score of the given mask is incremented by `1`.
+    ///
+    /// Behavior
+    /// - if sum participant has not submitted a mask yet, return Ok
+    /// - if pet protocol error, return [`MaskDictIncrError::`]
+    ///
+    /// # Errors:
+    ///
+    /// ## IO
+    /// - RedisError
+    ///
+    /// ## Protocol
+    /// - [`MaskDictIncrError::UnknownSumPk`]
+    /// - [`MaskDictIncrError::MaskAlreadySubmitted`]
+    async fn incr_mask_score(
+        &self,
+        pk: &SumParticipantPublicKey,
+        mask: &MaskObject,
+    ) -> StorageResult<MaskDictIncr>;
+
+    /// Gets the two masks with the highest score.
+    ///
+    /// Behavior
+    /// - if no masks exist, return Ok(empty vec) or None?
+    /// - if only one mask exists, return the mask
+    /// - if two masks exist with the same score, return both
+    /// - if two masks exist with the different score, return both (in descending order)
+    ///
+    /// # Errors:
+    ///
+    /// ## IO
+    /// - RedisError
+    ///
+    /// ## Protocol
+    /// None
+    async fn get_best_masks(&self) -> StorageResult<Vec<(MaskObject, u64)>>;
+
+    /// Gets the number of unique masks.
+    ///
+    /// # Errors:
+    ///
+    /// ## IO
+    /// - RedisError
+    ///
+    /// ## Protocol
+    /// None
+    async fn get_number_of_unique_masks(&self) -> StorageResult<u64>;
+
+    /////////// Data
+
+    /// Deletes all coordinator data in the current database.
+    ///
+    /// # Errors:
+    ///
+    /// ## IO
+    /// - RedisError
+    ///
+    /// ## Protocol
+    /// None
+    async fn delete_coordinator_data(&self) -> StorageResult<()>;
+
+    /// Deletes the [`SumDict`], [`SeedDict`] and mask dictionary.
+    ///
+    /// # Errors:
+    ///
+    /// ## IO
+    /// - RedisError
+    ///
+    /// ## Protocol
+    /// None
+    async fn delete_dicts(&self) -> StorageResult<()>;
+
+    /////////// Global model
+
+    /// Set a global model.
+    ///
+    /// Behavior
+    /// - if the global model already exists (same id),  ?
+    /// - if the global model does not exist,  write model and return id
+    ///
+    /// # Errors:
+    ///
+    /// ## IO
+    /// - S3Error
+    ///
+    /// ## Protocol
+    /// None
+    async fn set_global_model(
+        &self,
+        round_id: u64,
+        round_seed: &RoundSeed,
+        global_model: &Model,
+    ) -> Result<String, StorageError>;
+
+    /// Gets a global model.
+    ///
+    /// Behavior
+    /// - if the global model does not exist, return Ok(None)
+    /// - if the global model exists, return Ok(Some(Model))
+    ///
+    /// # Errors:
+    ///
+    /// ## IO
+    /// - S3Error
+    ///
+    /// ## Protocol
+    /// None
+    async fn get_global_model(&self, id: &str) -> StorageResult<Option<Model>>;
```

Comment:
I know this would incur some complexity but I'm wondering whether it would make sense to have separate traits for the persistent storage (s3) and volatile storage (redis). The problem with having both in the same trait is that will have to hardcode storage combinations. For instance the `ExternalStorage` below is for Redis + S3. But what if I want Redis + local filesystem? Then we'll have to re-implement the `Storage` trait and copy paste all the redis code.

Reply:
I agree I also had this idea. this way we allow different combinations. I will try to implement this